gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/DomainsResource.java

```diff
@@ -69,20 +78,41 @@
-                        .flatMapMaybe(domain -> hasPermission(authenticatedUser,
-                                or(of(ReferenceType.DOMAIN, domain.getId(), Permission.DOMAIN, Acl.READ),
-                                        of(ReferenceType.ENVIRONMENT, environmentId, Permission.DOMAIN, Acl.READ),
-                                        of(ReferenceType.ORGANIZATION, organizationId, Permission.DOMAIN, Acl.READ)))
-                                .filter(Boolean::booleanValue).map(permit -> domain))
-                        .map(this::filterDomainInfos)
-                        .sorted((o1, o2) -> String.CASE_INSENSITIVE_ORDER.compare(o1.getName(), o2.getName()))
-                        .toList())
+                .andThen(environmentService.findById(environmentId, organizationId)
+                        .flatMap(__ -> {
+                    if (query != null) {
+                        return domainService.search(environmentId, query)
+                                .flatMapMaybe(domain -> hasPermission(authenticatedUser,
+                                        or(of(ReferenceType.DOMAIN, domain.getId(), Permission.DOMAIN, Acl.READ),
+                                                of(ReferenceType.ENVIRONMENT, environmentId, Permission.DOMAIN, Acl.READ),
+                                                of(ReferenceType.ORGANIZATION, organizationId, Permission.DOMAIN, Acl.READ)))
+                                        .filter(Boolean::booleanValue).map(permit -> domain))
+                                .map(this::filterDomainInfos)
+                                .sorted((o1, o2) -> String.CASE_INSENSITIVE_ORDER.compare(o1.getName(), o2.getName()))
+                                .toList()
+                                .map(domains -> new Page<Domain>(domains.stream().skip((long) page * size).limit(size).collect(Collectors.toList()), page, domains.size()));
+                    } else {
+                        return domainService.findAllByEnvironment(organizationId, environmentId)
+                                .flatMapMaybe(domain -> hasPermission(authenticatedUser,
+                                        or(of(ReferenceType.DOMAIN, domain.getId(), Permission.DOMAIN, Acl.READ),
+                                                of(ReferenceType.ENVIRONMENT, environmentId, Permission.DOMAIN, Acl.READ),
+                                                of(ReferenceType.ORGANIZATION, organizationId, Permission.DOMAIN, Acl.READ)))
+                                        .filter(Boolean::booleanValue).map(permit -> domain))
+                                .map(this::filterDomainInfos)
+                                .sorted((o1, o2) -> String.CASE_INSENSITIVE_ORDER.compare(o1.getName(), o2.getName()))
+                                .toList()
+                                .map(domains -> new Page<Domain>(domains.stream().skip((long) page * size).limit(size).collect(Collectors.toList()), page, domains.size()));
```

Comment:
The flatMapMaybe method is duplicated for both search and findAllByEnvironment

I think we can do something like : 

```
.andThen(environmentService.findById(environmentId, organizationId)
  .flatMap(__ -> query != null ? domainService.search() : domain.findAllByEnv())
    .flatMapMaybe(domain -> hasPermission(authenticatedUser,
                                        or(of(ReferenceType.DOMAIN, domain.getId(), Permission.DOMAIN, Acl.READ),
                                                of(ReferenceType.ENVIRONMENT, environmentId, Permission.DOMAIN, Acl.READ),
                                                of(ReferenceType.ORGANIZATION, organizationId, Permission.DOMAIN, Acl.READ)))
                                        .filter(Boolean::booleanValue).map(permit -> domain))
                                .map(this::filterDomainInfos)
                                .sorted((o1, o2) -> String.CASE_INSENSITIVE_ORDER.compare(o1.getName(), o2.getName()))
                                .toList()
                                .map(domains -> new Page<Domain>(domains.stream().skip((long) page * size).limit(size).collect(Collectors.toList()), page, domains.size()));
  ```

Reply:
resolved